# Objective
Forecast sales (no. of items sold) for each item and store for the next 28 days

20200527: Diff from m5-forecasting_v1.ipynb: Here we do predictions for 1 state at a time

20200606: Diff from m5-forecasting_v2.ipynb: Here we do more feature engineering e.g. target_store, target_item, target_dept

20200607: Diff from m5-forecasting_v3_feature_eng.ipynb: Here we remove features that have an importance score of close to 0

In [93]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import gc
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import time

from matplotlib import pyplot as plt
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder
from xgboost import XGBRegressor

%matplotlib inline

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [94]:
##### Input params ########
state = 'CA'
store = 'CA_1' # CA_1, CA_2, CA_3, CA_4, TX_1, TX_2, TX_3, WI_1, WI_2, WI_3

model_seed = 102
sample_seed = 103              # for sampling the dataframe

n_estimators = 100             # for the initial model before tuning. default = 100
max_depth = 3                  # for the initial model before tuning. default = 3
learning_rate = 0.1            # for the initial model before tuning. default = 0.1
min_child_weight = 1           # for the initial model before tuning. default = 1
subsample = 1                  # for the initial model before tuning. default = 1
colsample_bytree = 1           # for the initial model before tuning. default = 1
colsample_bylevel = 1          # for the initial model before tuning. default = 1
train_test_split_seed = 111    # 111
model_seed = 100

train_start = 0
train_end = 1885               # use date_block_num from train_start to train_end as train set
val_date_block_num = 1886      # date_block_num for validation set
pred_start = 1886              # do pred starting from this day
pred_end = 1913                # do pred until and include this day

# train_start = 0
# train_end = 1913               # use date_block_num from train_start to train_end as train set
# val_date_block_num = 1913      # date_block_num for validation set
# pred_start = 1914            # do pred starting from this day
# pred_end = 1941              # do pred until and include this day

shift_range = [1, 2, 3, 4, 5, 6, 7, 30, 365] # Use values from last 7 days, from 30 days ago, and from 365 days ago

fontsize = 14
ticklabelsize = 14
###########################

In [95]:
ticStart = time.time()

# Common functions

In [96]:
def downcast_dtypes(df):
    '''
        Changes column types in the dataframe: 
                
                `float64` type to `float32`
                `int64`   type to `int16`
    '''
    
    # Select columns to downcast
    float_cols = [c for c in df if df[c].dtype == "float64"]
    int_cols =   [c for c in df if df[c].dtype == "int64"]
    
    # Downcast
    df[float_cols] = df[float_cols].astype(np.float32)
    df[int_cols]   = df[int_cols].astype(np.int16)
    
    return df

def get_lags(index_cols, cols_to_rename, all_data, shift_range):
    """
    For a dataframe all_data, get the lags specified in cols_to_rename
    Input
        index_cols    : e.g. ['item_id', 'store_id', 'date_block_num']. These are the cols to do merging with the
                        lagged dataset
        cols_to_rename: e.g. ['target']. Based on this example, all_data will be returned with columns target_lag_1, 
                        target_lag_2, ..., target_lag_365.
        all_data      : need to have a column 'date_block_num' to indicate chronological order
        shift_range   : e.g. [1, 2, 3, 4, 5, 6, 7, 30, 365]
    Output
        all_data
    """
    for day_shift in shift_range:
        train_shift = all_data[index_cols + cols_to_rename].copy()
    
        # E.g. variable of 0 becomes 1, for day_shift = 1.
        # So when this is merged with variable of 1 in all_data, this will represent lag of 1.
        train_shift['date_block_num'] = train_shift['date_block_num'] + day_shift
    
        foo = lambda x: '{}_lag_{}'.format(x, day_shift) if x in cols_to_rename else x
        train_shift = train_shift.rename(columns=foo)

        all_data = pd.merge(all_data, train_shift, on=index_cols, how='left')
    
        # Fill the NaNs with 0
        for x in cols_to_rename:
            all_data[x + '_lag_' + str(day_shift)].fillna(0, inplace=True)
    
    del train_shift
    gc.collect();

    return all_data


In [97]:
def get_agg_features(all_data):
    """
    Do feature engineering to get 3 features for all_data: target_store, target_item, target_dept
    """
    # Create (shop_id, date_block_num) aggregates
    # This gives the number of items each shop sold per date_block_num as a feature
    # We won't have the above for the test set, so can only use its lagged features 
    gb = all_data.groupby(['store_id', 'date_block_num'], as_index=False).agg({'target':[('target_store','sum')]})
    gb.columns = [col[0] if col[-1]=='' else col[-1] for col in gb.columns.values]
    all_data = all_data.merge(gb, 
                              left_on=['store_id', 'date_block_num'], 
                              right_on=['store_id', 'date_block_num'],
                              how='left').fillna(0)
    
    # Create (item_id, date_block_num) aggregates
    # This gives the number of items sold per item_id per date_block_num as a feature
    # We won't have the above for the test set, so can only use its lagged features 
    gb = all_data.groupby(['item_id', 'date_block_num'], as_index=False).agg({'target':[('target_item','sum')]})
    gb.columns = [col[0] if col[-1]=='' else col[-1] for col in gb.columns.values]
    all_data = all_data.merge(gb, 
                              left_on=['item_id', 'date_block_num'], 
                              right_on=['item_id', 'date_block_num'],
                              how='left').fillna(0)
    
    # Create (dept_id, date_block_num) aggregates
    # This gives the number of items sold per dept_id per date_block_num as a feature
    # We won't have the above for the test set, so can only use its lagged features 
    gb = all_data.groupby(['dept_id', 'date_block_num'], as_index=False).agg({'target':[('target_dept','sum')]})
    gb.columns = [col[0] if col[-1]=='' else col[-1] for col in gb.columns.values]
    all_data = all_data.merge(gb, 
                              left_on=['dept_id', 'date_block_num'], 
                              right_on=['dept_id', 'date_block_num'],
                              how='left').fillna(0)
    
    return all_data

# Load data

In [98]:
sell_prices = pd.read_csv("./data/sell_prices.csv")
sell_prices.head()

store_id        item_id  wm_yr_wk  sell_price
0     CA_1  HOBBIES_1_001     11325        9.58
1     CA_1  HOBBIES_1_001     11326        9.58
2     CA_1  HOBBIES_1_001     11327        8.26
3     CA_1  HOBBIES_1_001     11328        8.26
4     CA_1  HOBBIES_1_001     11329        8.26

File 2: “sell_prices.csv”

Contains information about the price of the products sold per store and date.
* store_id: The id of the store where the product is sold. 
* item_id: The id of the product.
* wm_yr_wk: The id of the week.
* sell_price: The price of the product for the given week/store. The price is provided per week (average across seven days). If not available, this means that the product was not sold during the examined week. Note that although prices are constant at weekly basis, they may change through time (both training and test set).

In [99]:
calendar = pd.read_csv("./data/calendar.csv")
calendar

date  wm_yr_wk    weekday  wday  month  year       d  \
0     2011-01-29     11101   Saturday     1      1  2011     d_1   
1     2011-01-30     11101     Sunday     2      1  2011     d_2   
2     2011-01-31     11101     Monday     3      1  2011     d_3   
3     2011-02-01     11101    Tuesday     4      2  2011     d_4   
4     2011-02-02     11101  Wednesday     5      2  2011     d_5   
5     2011-02-03     11101   Thursday     6      2  2011     d_6   
6     2011-02-04     11101     Friday     7      2  2011     d_7   
7     2011-02-05     11102   Saturday     1      2  2011     d_8   
8     2011-02-06     11102     Sunday     2      2  2011     d_9   
9     2011-02-07     11102     Monday     3      2  2011    d_10   
10    2011-02-08     11102    Tuesday     4      2  2011    d_11   
11    2011-02-09     11102  Wednesday     5      2  2011    d_12   
12    2011-02-10     11102   Thursday     6      2  2011    d_13   
13    2011-02-11     11102     Friday     7      2  2011    d_14   
14    2011-02-12     11103   Saturday     1      2  2011    d_15   
15    2011-02-13     11103     Sunday     2      2  2011    d_16   
16    2011-02-14     11103     Monday     3      2  2011    d_17   
17    2011-02-15     11103    Tuesday     4      2  2011    d_18   
18    2011-02-16     11103  Wednesday     5      2  2011    d_19   
19    2011-02-17     11103   Thursday     6      2  2011    d_20   
20    2011-02-18     11103     Friday     7      2  2011    d_21   
21    2011-02-19     11104   Saturday     1      2  2011    d_22   
22    2011-02-20     11104     Sunday     2      2  2011    d_23   
23    2011-02-21     11104     Monday     3      2  2011    d_24   
24    2011-02-22     11104    Tuesday     4      2  2011    d_25   
25    2011-02-23     11104  Wednesday     5      2  2011    d_26   
26    2011-02-24     11104   Thursday     6      2  2011    d_27   
27    2011-02-25     11104     Friday     7      2  2011    d_28   
28    2011-02-26     11105   Saturday     1      2  2011    d_29   
29    2011-02-27     11105     Sunday     2      2  2011    d_30   
...          ...       ...        ...   ...    ...   ...     ...   
1939  2016-05-21     11617   Saturday     1      5  2016  d_1940   
1940  2016-05-22     11617     Sunday     2      5  2016  d_1941   
1941  2016-05-23     11617     Monday     3      5  2016  d_1942   
1942  2016-05-24     11617    Tuesday     4      5  2016  d_1943   
1943  2016-05-25     11617  Wednesday     5      5  2016  d_1944   
1944  2016-05-26     11617   Thursday     6      5  2016  d_1945   
1945  2016-05-27     11617     Friday     7      5  2016  d_1946   
1946  2016-05-28     11618   Saturday     1      5  2016  d_1947   
1947  2016-05-29     11618     Sunday     2      5  2016  d_1948   
1948  2016-05-30     11618     Monday     3      5  2016  d_1949   
1949  2016-05-31     11618    Tuesday     4      5  2016  d_1950   
1950  2016-06-01     11618  Wednesday     5      6  2016  d_1951   
1951  2016-06-02     11618   Thursday     6      6  2016  d_1952   
1952  2016-06-03     11618     Friday     7      6  2016  d_1953   
1953  2016-06-04     11619   Saturday     1      6  2016  d_1954   
1954  2016-06-05     11619     Sunday     2      6  2016  d_1955   
1955  2016-06-06     11619     Monday     3      6  2016  d_1956   
1956  2016-06-07     11619    Tuesday     4      6  2016  d_1957   
1957  2016-06-08     11619  Wednesday     5      6  2016  d_1958   
1958  2016-06-09     11619   Thursday     6      6  2016  d_1959   
1959  2016-06-10     11619     Friday     7      6  2016  d_1960   
1960  2016-06-11     11620   Saturday     1      6  2016  d_1961   
1961  2016-06-12     11620     Sunday     2      6  2016  d_1962   
1962  2016-06-13     11620     Monday     3      6  2016  d_1963   
1963  2016-06-14     11620    Tuesday     4      6  2016  d_1964   
1964  2016-06-15     11620  Wednesday     5      6  2016  d_1965   
1965  2016-06-16     11620   Thursday     6      6  2016  d_1966   


File 1: “calendar.csv” 

Contains information about the dates the products are sold.
* date: The date in a “y-m-d” format.
* wm_yr_wk: The id of the week the date belongs to.
* weekday: The type of the day (Saturday, Sunday, …, Friday).
* wday: The id of the weekday, starting from Saturday.
* month: The month of the date.
* year: The year of the date.
* event_name_1: If the date includes an event, the name of this event.
* event_type_1: If the date includes an event, the type of this event.
* event_name_2: If the date includes a second event, the name of this event.
* event_type_2: If the date includes a second event, the type of this event.
* snap_CA, snap_TX, and snap_WI: A binary variable (0 or 1) indicating whether the stores of CA, TX or WI allow SNAP purchases on the examined date. 1 indicates that SNAP purchases are allowed.


In [100]:
sales_tr_val = pd.read_csv("./data/sales_train_validation.csv")
print("len(sales_tr_val) = " + str(len(sales_tr_val)))
sales_tr_val.head()

len(sales_tr_val) = 30490


id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  d_1  d_2  d_3  d_4  ...  d_1904  d_1905  d_1906  d_1907  d_1908  \
0       CA    0    0    0    0  ...       1       3       0       1       1   
1       CA    0    0    0    0  ...       0       0       0       0       0   
2       CA    0    0    0    0  ...       2       1       2       1       1   
3       CA    0    0    0    0  ...       1       0       5       4       1   
4       CA    0    0    0    0  ...       2       1       1       0       1   

   d_1909  d_1910  d_1911  d_1912  d_1913  
0       1       3       0       1       1  
1       1       0       0       0       0  
2       1       0       1       1       1  
3       0       1       3       7       2  
4       1       2       2       2       4  

[5 rows x 1919 columns]

For each (item_id, store_id), we have the number of items sold for 1913 days or 5.24 years.

There are 30,490 distinct id's, and 1913 days which gives 30,490*1913 = 58,327,370

In [101]:
# Keep only data from one state
sales_tr_val = sales_tr_val[(sales_tr_val['state_id']==state) & (sales_tr_val['store_id']==store)]
print("len(sales_tr_val) = " + str(len(sales_tr_val)))

len(sales_tr_val) = 3049


In [102]:
sample_sub = pd.read_csv("./data/sample_submission.csv")
print(len(sample_sub))
sample_sub.head()

60980


id  F1  F2  F3  F4  F5  F6  F7  F8  F9  ...  \
0  HOBBIES_1_001_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
1  HOBBIES_1_002_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
2  HOBBIES_1_003_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
3  HOBBIES_1_004_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
4  HOBBIES_1_005_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   

   F19  F20  F21  F22  F23  F24  F25  F26  F27  F28  
0    0    0    0    0    0    0    0    0    0    0  
1    0    0    0    0    0    0    0    0    0    0  
2    0    0    0    0    0    0    0    0    0    0  
3    0    0    0    0    0    0    0    0    0    0  
4    0    0    0    0    0    0    0    0    0    0  

[5 rows x 29 columns]

Provide the forecast for each id for the next 28 days. For the validation rows, this corresponds to d_1914 - d_1941, and for the evaluation rows, this corresponds to d_1942 - d_1969. (Note: a month before the competition close, the ground truth for the validation rows will be provided.)

Note 30,490*2 = 60,980 which is the number of rows in the submission file.

# Pre-process data

In [103]:
print(calendar.memory_usage().sum())

# # In calendar 'd' column, change 'd_1' to 1
# calendar['d'] = calendar.apply(lambda row: row['d'].split('_')[1], axis=1)

# Replance NaN with nil
calendar = calendar.fillna(value = 'nil')

# Downcast dtypes from 64 bit to save memory
calendar= downcast_dtypes(calendar)

# Change to category dtype
calendar[["year", "event_name_1", "event_type_1", "event_name_2", "event_type_2"]] = \
    calendar[["year", "event_name_1", "event_type_1", "event_name_2", "event_type_2"]] .astype("category")

# Drop weekday column since its not used
calendar.drop(['weekday'], axis=1, inplace=True)

print(calendar.memory_usage().sum())

calendar.head()

220608
67297


date  wm_yr_wk  wday  month  year    d event_name_1 event_type_1  \
0  2011-01-29     11101     1      1  2011  d_1          nil          nil   
1  2011-01-30     11101     2      1  2011  d_2          nil          nil   
2  2011-01-31     11101     3      1  2011  d_3          nil          nil   
3  2011-02-01     11101     4      2  2011  d_4          nil          nil   
4  2011-02-02     11101     5      2  2011  d_5          nil          nil   

  event_name_2 event_type_2  snap_CA  snap_TX  snap_WI  
0          nil          nil        0        0        0  
1          nil          nil        0        0        0  
2          nil          nil        0        0        0  
3          nil          nil        1        1        0  
4          nil          nil        1        0        1

In [104]:
# Downcast dtypes from 64 bit to save memory
print(sell_prices.memory_usage().sum())
sell_prices = downcast_dtypes(sell_prices)

sell_prices[["store_id", "item_id"]] = \
    sell_prices[["store_id","item_id"]].astype("category")

print(sell_prices.memory_usage().sum())

218915952
61676881


In [105]:
# Downcast dtypes from 64 bit to save memory
print(sales_tr_val.memory_usage().sum())
sales_tr_val = downcast_dtypes(sales_tr_val)

sales_tr_val[["item_id", "dept_id", "cat_id", "store_id", "state_id"]] = \
    sales_tr_val[["item_id", "dept_id", "cat_id", "store_id", "state_id"]].astype("category")

print(sales_tr_val.memory_usage().sum())

46832640
11839520


In [106]:
# Melt sales_tr_val to switch from a wide to a long dataframe
tic = time.time()
value_vars = ['d_' + str(x) for x in range(1, 1914)]
sale_tr_val_melt = pd.melt(sales_tr_val, 
                           id_vars=['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id'], 
                           value_vars=value_vars)

toc = time.time()
print("Time taken = " + str(toc-tic) + ' s')
sale_tr_val_melt.head()

Time taken = 12.562716245651245 s


id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id variable  value  
0       CA      d_1      0  
1       CA      d_1      0  
2       CA      d_1      0  
3       CA      d_1      0  
4       CA      d_1      0

In [107]:
del sales_tr_val
gc.collect();
sales_tr_val = pd.DataFrame()

In [108]:
# Merge with calendar to get events info
sale_tr_val_melt_merged = sale_tr_val_melt.merge(calendar, 
                                                 left_on='variable', 
                                                 right_on='d', 
                                                 how='left')

# Drop unneccesary columns
sale_tr_val_melt_merged.drop(columns=['d'], inplace=True)

sale_tr_val_melt_merged.head()

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id variable  value        date  wm_yr_wk  wday  month  year  \
0       CA      d_1      0  2011-01-29     11101     1      1  2011   
1       CA      d_1      0  2011-01-29     11101     1      1  2011   
2       CA      d_1      0  2011-01-29     11101     1      1  2011   
3       CA      d_1      0  2011-01-29     11101     1      1  2011   
4       CA      d_1      0  2011-01-29     11101     1      1  2011   

  event_name_1 event_type_1 event_name_2 event_type_2  snap_CA  snap_TX  \
0          nil          nil          nil          nil        0        0   
1          nil          nil          nil          nil        0        0   
2          nil          nil          nil          nil        0        0   
3          nil          nil          nil          nil        0        0   
4          nil          nil          nil          nil        0        0   

   snap_WI  
0        0  
1        0  
2        0  
3        0  
4        0

In [109]:
# del calendar
# gc.collect();
# calendar = pd.DataFrame()

# del sale_tr_val_melt
# gc.collect();
# sale_tr_val_melt = pd.DataFrame()

In [110]:
# Merge with sell_prices to get price info
sale_tr_val_melt_merged2 = sale_tr_val_melt_merged.merge(sell_prices, 
                                                         left_on=['store_id', 'item_id', 'wm_yr_wk'], 
                                                         right_on=['store_id', 'item_id', 'wm_yr_wk'], 
                                                         how='left')
sale_tr_val_melt_merged2.head()

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id variable  value        date  wm_yr_wk  ...  month  year  \
0       CA      d_1      0  2011-01-29     11101  ...      1  2011   
1       CA      d_1      0  2011-01-29     11101  ...      1  2011   
2       CA      d_1      0  2011-01-29     11101  ...      1  2011   
3       CA      d_1      0  2011-01-29     11101  ...      1  2011   
4       CA      d_1      0  2011-01-29     11101  ...      1  2011   

  event_name_1 event_type_1 event_name_2 event_type_2 snap_CA  snap_TX  \
0          nil          nil          nil          nil       0        0   
1          nil          nil          nil          nil       0        0   
2          nil          nil          nil          nil       0        0   
3          nil          nil          nil          nil       0        0   
4          nil          nil          nil          nil       0        0   

   snap_WI  sell_price  
0        0         NaN  
1        0         NaN  
2        0         NaN  
3        0         NaN  
4        0         NaN  

[5 rows x 21 columns]

In [111]:
# del sell_prices
# gc.collect();
# sell_prices = pd.DataFrame()

del sale_tr_val_melt_merged
gc.collect();
sale_tr_val_melt_merged = pd.DataFrame()

In [112]:
# Get no. of distinct values in each column
print("item_id uniques = " + str(sale_tr_val_melt_merged2['item_id'].nunique()))
print("dept_id uniques = " + str(sale_tr_val_melt_merged2['dept_id'].nunique()))
print("cat_id uniques = " + str(sale_tr_val_melt_merged2['cat_id'].nunique()))
print("store_id uniques = " + str(sale_tr_val_melt_merged2['store_id'].nunique()))
print("state_id uniques = " + str(sale_tr_val_melt_merged2['state_id'].nunique()))

item_id uniques = 3049
dept_id uniques = 7
cat_id uniques = 3
store_id uniques = 1
state_id uniques = 1


In [113]:
print(sale_tr_val_melt_merged2['dept_id'].unique())

[HOBBIES_1, HOBBIES_2, HOUSEHOLD_1, HOUSEHOLD_2, FOODS_1, FOODS_2, FOODS_3]
Categories (7, object): [HOBBIES_1, HOBBIES_2, HOUSEHOLD_1, HOUSEHOLD_2, FOODS_1, FOODS_2, FOODS_3]


In [114]:
print(sale_tr_val_melt_merged2['cat_id'].unique())

[HOBBIES, HOUSEHOLD, FOODS]
Categories (3, object): [HOBBIES, HOUSEHOLD, FOODS]


In [115]:
print(sale_tr_val_melt_merged2['store_id'].unique())

['CA_1']


In [116]:
print(sale_tr_val_melt_merged2['state_id'].unique())

[CA]
Categories (1, object): [CA]


In [117]:
# Convert variable to a numerical column ie. d_1 to 1
sale_tr_val_melt_merged2['variable'] = sale_tr_val_melt_merged2['variable'].str.split('_').str[1]
sale_tr_val_melt_merged2

id        item_id    dept_id   cat_id  \
0        HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES   
1        HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES   
2        HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES   
3        HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES   
4        HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES   
5        HOBBIES_1_006_CA_1_validation  HOBBIES_1_006  HOBBIES_1  HOBBIES   
6        HOBBIES_1_007_CA_1_validation  HOBBIES_1_007  HOBBIES_1  HOBBIES   
7        HOBBIES_1_008_CA_1_validation  HOBBIES_1_008  HOBBIES_1  HOBBIES   
8        HOBBIES_1_009_CA_1_validation  HOBBIES_1_009  HOBBIES_1  HOBBIES   
9        HOBBIES_1_010_CA_1_validation  HOBBIES_1_010  HOBBIES_1  HOBBIES   
10       HOBBIES_1_011_CA_1_validation  HOBBIES_1_011  HOBBIES_1  HOBBIES   
11       HOBBIES_1_012_CA_1_validation  HOBBIES_1_012  HOBBIES_1  HOBBIES   
12       HOBBIES_1_013_CA_1_validation  HOBBIES_1_013  HOBBIES_1  HOBBIES   
13       HOBBIES_1_014_CA_1_validation  HOBBIES_1_014  HOBBIES_1  HOBBIES   
14       HOBBIES_1_015_CA_1_validation  HOBBIES_1_015  HOBBIES_1  HOBBIES   
15       HOBBIES_1_016_CA_1_validation  HOBBIES_1_016  HOBBIES_1  HOBBIES   
16       HOBBIES_1_017_CA_1_validation  HOBBIES_1_017  HOBBIES_1  HOBBIES   
17       HOBBIES_1_018_CA_1_validation  HOBBIES_1_018  HOBBIES_1  HOBBIES   
18       HOBBIES_1_019_CA_1_validation  HOBBIES_1_019  HOBBIES_1  HOBBIES   
19       HOBBIES_1_020_CA_1_validation  HOBBIES_1_020  HOBBIES_1  HOBBIES   
20       HOBBIES_1_021_CA_1_validation  HOBBIES_1_021  HOBBIES_1  HOBBIES   
21       HOBBIES_1_022_CA_1_validation  HOBBIES_1_022  HOBBIES_1  HOBBIES   
22       HOBBIES_1_023_CA_1_validation  HOBBIES_1_023  HOBBIES_1  HOBBIES   
23       HOBBIES_1_024_CA_1_validation  HOBBIES_1_024  HOBBIES_1  HOBBIES   
24       HOBBIES_1_025_CA_1_validation  HOBBIES_1_025  HOBBIES_1  HOBBIES   
25       HOBBIES_1_026_CA_1_validation  HOBBIES_1_026  HOBBIES_1  HOBBIES   
26       HOBBIES_1_027_CA_1_validation  HOBBIES_1_027  HOBBIES_1  HOBBIES   
27       HOBBIES_1_028_CA_1_validation  HOBBIES_1_028  HOBBIES_1  HOBBIES   
28       HOBBIES_1_029_CA_1_validation  HOBBIES_1_029  HOBBIES_1  HOBBIES   
29       HOBBIES_1_030_CA_1_validation  HOBBIES_1_030  HOBBIES_1  HOBBIES   
...                                ...            ...        ...      ...   
5832707    FOODS_3_798_CA_1_validation    FOODS_3_798    FOODS_3    FOODS   
5832708    FOODS_3_799_CA_1_validation    FOODS_3_799    FOODS_3    FOODS   
5832709    FOODS_3_800_CA_1_validation    FOODS_3_800    FOODS_3    FOODS   
5832710    FOODS_3_801_CA_1_validation    FOODS_3_801    FOODS_3    FOODS   
5832711    FOODS_3_802_CA_1_validation    FOODS_3_802    FOODS_3    FOODS   
5832712    FOODS_3_803_CA_1_validation    FOODS_3_803    FOODS_3    FOODS   
5832713    FOODS_3_804_CA_1_validation    FOODS_3_804    FOODS_3    FOODS   
5832714    FOODS_3_805_CA_1_validation    FOODS_3_805    FOODS_3    FOODS   
5832715    FOODS_3_806_CA_1_validation    FOODS_3_806    FOODS_3    FOODS   
5832716    FOODS_3_807_CA_1_validation    FOODS_3_807    FOODS_3    FOODS   
5832717    FOODS_3_808_CA_1_validation    FOODS_3_808    FOODS_3    FOODS   
5832718    FOODS_3_809_CA_1_validation    FOODS_3_809    FOODS_3    FOODS   
5832719    FOODS_3_810_CA_1_validation    FOODS_3_810    FOODS_3    FOODS   
5832720    FOODS_3_811_CA_1_validation    FOODS_3_811    FOODS_3    FOODS   
5832721    FOODS_3_812_CA_1_validation    FOODS_3_812    FOODS_3    FOODS   
5832722    FOODS_3_813_CA_1_validation    FOODS_3_813    FOODS_3    FOODS   
5832723    FOODS_3_814_CA_1_validation    FOODS_3_814    FOODS_3    FOODS   
5832724    FOODS_3_815_CA_1_validation    FOODS_3_815    FOODS_3    FOODS   
5832725    FOODS_3_816_CA_1_validation    FOODS_3_816    FOODS_3    FOODS   
5832726    FOODS_3_817_CA_1_validation    FOODS_3_817    FOODS_3    FOODS   
5832727    FOODS_3_818_CA_1_vali

In [118]:
# Convert variable to int
sale_tr_val_melt_merged2['variable'] = sale_tr_val_melt_merged2['variable'].astype('int16')

# Create features

In [119]:
all_data = sale_tr_val_melt_merged2
all_data = all_data.rename(columns={'variable': 'date_block_num', 
                                    'value': 'target'})

In [120]:
# Remove unnecessaray column
all_data.drop(['date'], axis=1, inplace=True)

In [121]:
# Convert features to categorical
columns_to_be_labeled = [
    'year',
    'event_name_1',
    'event_type_1',
    'event_name_2',
    'event_type_2',
    'wm_yr_wk',
    'item_id',
    'store_id',
    'dept_id',
    'cat_id',
    'state_id'
]
for column in columns_to_be_labeled:
    all_data.loc[:, column] = LabelEncoder().fit_transform(all_data[column]) # fit_transform: Fit label encoder and return encoded labels. Note the encoded labels are integers!!

all_data.head()

id  item_id  dept_id  cat_id  store_id  \
0  HOBBIES_1_001_CA_1_validation     1437        3       1         0   
1  HOBBIES_1_002_CA_1_validation     1438        3       1         0   
2  HOBBIES_1_003_CA_1_validation     1439        3       1         0   
3  HOBBIES_1_004_CA_1_validation     1440        3       1         0   
4  HOBBIES_1_005_CA_1_validation     1441        3       1         0   

   state_id  date_block_num  target  wm_yr_wk  wday  month  year  \
0         0               1       0         0     1      1     0   
1         0               1       0         0     1      1     0   
2         0               1       0         0     1      1     0   
3         0               1       0         0     1      1     0   
4         0               1       0         0     1      1     0   

   event_name_1  event_type_1  event_name_2  event_type_2  snap_CA  snap_TX  \
0            30             4             4             2        0        0   
1            30             4             4             2        0        0   
2            30             4             4             2        0        0   
3            30             4             4             2        0        0   
4            30             4             4             2        0        0   

   snap_WI  sell_price  
0        0         NaN  
1        0         NaN  
2        0         NaN  
3        0         NaN  
4        0         NaN

In [122]:
# Create (shop_id, date_block_num) aggregates
# This gives the number of items each shop sold per date_block_num as a feature
# We won't have the above for the test set, so can only use its lagged features 
gb = all_data.groupby(['store_id', 'date_block_num'], as_index=False).agg({'target':{'target_store':'sum'}})
gb.columns = [col[0] if col[-1]=='' else col[-1] for col in gb.columns.values]
all_data = all_data.merge(gb, 
                          left_on=['store_id', 'date_block_num'], 
                          right_on=['store_id', 'date_block_num'],
                          how='left').fillna(0)
all_data.head()

/Users/yibin/anaconda3/lib/python3.7/site-packages/pandas/core/groupby/generic.py:1315: FutureWarning: using a dict with renaming is deprecated and will be removed in a future version
  return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)


id  item_id  dept_id  cat_id  store_id  \
0  HOBBIES_1_001_CA_1_validation     1437        3       1         0   
1  HOBBIES_1_002_CA_1_validation     1438        3       1         0   
2  HOBBIES_1_003_CA_1_validation     1439        3       1         0   
3  HOBBIES_1_004_CA_1_validation     1440        3       1         0   
4  HOBBIES_1_005_CA_1_validation     1441        3       1         0   

   state_id  date_block_num  target  wm_yr_wk  wday  ...  year  event_name_1  \
0         0               1       0         0     1  ...     0            30   
1         0               1       0         0     1  ...     0            30   
2         0               1       0         0     1  ...     0            30   
3         0               1       0         0     1  ...     0            30   
4         0               1       0         0     1  ...     0            30   

   event_type_1  event_name_2  event_type_2  snap_CA  snap_TX  snap_WI  \
0             4             4             2        0        0        0   
1             4             4             2        0        0        0   
2             4             4             2        0        0        0   
3             4             4             2        0        0        0   
4             4             4             2        0        0        0   

   sell_price  target_store  
0         0.0          4337  
1         0.0          4337  
2         0.0          4337  
3         0.0          4337  
4         0.0          4337  

[5 rows x 21 columns]

In [123]:
# Create (item_id, date_block_num) aggregates
# This gives the number of items sold per item_id per date_block_num as a feature
# We won't have the above for the test set, so can only use its lagged features 
gb = all_data.groupby(['item_id', 'date_block_num'], as_index=False).agg({'target':{'target_item':'sum'}})
gb.columns = [col[0] if col[-1]=='' else col[-1] for col in gb.columns.values]
all_data = all_data.merge(gb, 
                          left_on=['item_id', 'date_block_num'], 
                          right_on=['item_id', 'date_block_num'],
                          how='left').fillna(0)
all_data.head()

id  item_id  dept_id  cat_id  store_id  \
0  HOBBIES_1_001_CA_1_validation     1437        3       1         0   
1  HOBBIES_1_002_CA_1_validation     1438        3       1         0   
2  HOBBIES_1_003_CA_1_validation     1439        3       1         0   
3  HOBBIES_1_004_CA_1_validation     1440        3       1         0   
4  HOBBIES_1_005_CA_1_validation     1441        3       1         0   

   state_id  date_block_num  target  wm_yr_wk  wday  ...  event_name_1  \
0         0               1       0         0     1  ...            30   
1         0               1       0         0     1  ...            30   
2         0               1       0         0     1  ...            30   
3         0               1       0         0     1  ...            30   
4         0               1       0         0     1  ...            30   

   event_type_1  event_name_2  event_type_2  snap_CA  snap_TX  snap_WI  \
0             4             4             2        0        0        0   
1             4             4             2        0        0        0   
2             4             4             2        0        0        0   
3             4             4             2        0        0        0   
4             4             4             2        0        0        0   

   sell_price  target_store  target_item  
0         0.0          4337            0  
1         0.0          4337            0  
2         0.0          4337            0  
3         0.0          4337            0  
4         0.0          4337            0  

[5 rows x 22 columns]

In [124]:
# Create (dept_id, date_block_num) aggregates
# This gives the number of items sold per dept_id per date_block_num as a feature
# We won't have the above for the test set, so can only use its lagged features 
gb = all_data.groupby(['dept_id', 'date_block_num'], as_index=False).agg({'target':{'target_dept':'sum'}})
gb.columns = [col[0] if col[-1]=='' else col[-1] for col in gb.columns.values]
all_data = all_data.merge(gb, 
                          left_on=['dept_id', 'date_block_num'], 
                          right_on=['dept_id', 'date_block_num'],
                          how='left').fillna(0)
all_data.head()

id  item_id  dept_id  cat_id  store_id  \
0  HOBBIES_1_001_CA_1_validation     1437        3       1         0   
1  HOBBIES_1_002_CA_1_validation     1438        3       1         0   
2  HOBBIES_1_003_CA_1_validation     1439        3       1         0   
3  HOBBIES_1_004_CA_1_validation     1440        3       1         0   
4  HOBBIES_1_005_CA_1_validation     1441        3       1         0   

   state_id  date_block_num  target  wm_yr_wk  wday  ...  event_type_1  \
0         0               1       0         0     1  ...             4   
1         0               1       0         0     1  ...             4   
2         0               1       0         0     1  ...             4   
3         0               1       0         0     1  ...             4   
4         0               1       0         0     1  ...             4   

   event_name_2  event_type_2  snap_CA  snap_TX  snap_WI  sell_price  \
0             4             2        0        0        0         0.0   
1             4             2        0        0        0         0.0   
2             4             2        0        0        0         0.0   
3             4             2        0        0        0         0.0   
4             4             2        0        0        0         0.0   

   target_store  target_item  target_dept  
0          4337            0          528  
1          4337            0          528  
2          4337            0          528  
3          4337            0          528  
4          4337            0          528  

[5 rows x 23 columns]

In [125]:
# Create lag features
index_cols = ['item_id', 'store_id', 'date_block_num']
cols_to_rename = ['target', 'target_store', 'target_item', 'target_dept']

for day_shift in shift_range:
    print(day_shift)
    train_shift = all_data[index_cols + cols_to_rename].copy()
    
    # E.g. variable of 0 becomes 1, for day_shift = 1.
    # So when this is merged with variable of 1 in all_data, this will represent lag of 1.
    train_shift['date_block_num'] = train_shift['date_block_num'] + day_shift
    
    foo = lambda x: '{}_lag_{}'.format(x, day_shift) if x in cols_to_rename else x
    train_shift = train_shift.rename(columns=foo)

    all_data = pd.merge(all_data, train_shift, on=index_cols, how='left')
    
    # Fill the NaNs with 0
    for x in cols_to_rename:
        all_data[x + '_lag_' + str(day_shift)].fillna(0, inplace=True)
    
del train_shift
gc.collect();

all_data

1
2
3
4
5
6
7
30
365


id  item_id  dept_id  cat_id  store_id  \
0        HOBBIES_1_001_CA_1_validation     1437        3       1         0   
1        HOBBIES_1_002_CA_1_validation     1438        3       1         0   
2        HOBBIES_1_003_CA_1_validation     1439        3       1         0   
3        HOBBIES_1_004_CA_1_validation     1440        3       1         0   
4        HOBBIES_1_005_CA_1_validation     1441        3       1         0   
5        HOBBIES_1_006_CA_1_validation     1442        3       1         0   
6        HOBBIES_1_007_CA_1_validation     1443        3       1         0   
7        HOBBIES_1_008_CA_1_validation     1444        3       1         0   
8        HOBBIES_1_009_CA_1_validation     1445        3       1         0   
9        HOBBIES_1_010_CA_1_validation     1446        3       1         0   
10       HOBBIES_1_011_CA_1_validation     1447        3       1         0   
11       HOBBIES_1_012_CA_1_validation     1448        3       1         0   
12       HOBBIES_1_013_CA_1_validation     1449        3       1         0   
13       HOBBIES_1_014_CA_1_validation     1450        3       1         0   
14       HOBBIES_1_015_CA_1_validation     1451        3       1         0   
15       HOBBIES_1_016_CA_1_validation     1452        3       1         0   
16       HOBBIES_1_017_CA_1_validation     1453        3       1         0   
17       HOBBIES_1_018_CA_1_validation     1454        3       1         0   
18       HOBBIES_1_019_CA_1_validation     1455        3       1         0   
19       HOBBIES_1_020_CA_1_validation     1456        3       1         0   
20       HOBBIES_1_021_CA_1_validation     1457        3       1         0   
21       HOBBIES_1_022_CA_1_validation     1458        3       1         0   
22       HOBBIES_1_023_CA_1_validation     1459        3       1         0   
23       HOBBIES_1_024_CA_1_validation     1460        3       1         0   
24       HOBBIES_1_025_CA_1_validation     1461        3       1         0   
25       HOBBIES_1_026_CA_1_validation     1462        3       1         0   
26       HOBBIES_1_027_CA_1_validation     1463        3       1         0   
27       HOBBIES_1_028_CA_1_validation     1464        3       1         0   
28       HOBBIES_1_029_CA_1_validation     1465        3       1         0   
29       HOBBIES_1_030_CA_1_validation     1466        3       1         0   
...                                ...      ...      ...     ...       ...   
5832707    FOODS_3_798_CA_1_validation     1407        2       0         0   
5832708    FOODS_3_799_CA_1_validation     1408        2       0         0   
5832709    FOODS_3_800_CA_1_validation     1409        2       0         0   
5832710    FOODS_3_801_CA_1_validation     1410        2       0         0   
5832711    FOODS_3_802_CA_1_validation     1411        2       0         0   
5832712    FOODS_3_803_CA_1_validation     1412        2       0         0   
5832713    FOODS_3_804_CA_1_validation     1413        2       0         0   
5832714    FOODS_3_805_CA_1_validation     1414        2       0         0   
5832715    FOODS_3_806_CA_1_validation     1415        2       0         0   
5832716    FOODS_3_807_CA_1_validation     1416        2       0         0   
5832717    FOODS_3_808_CA_1_validation     1417        2       0         0   
5832718    FOODS_3_809_CA_1_validation     1418        2       0         0   
5832719    FOODS_3_810_CA_1_validation     1419        2       0         0   
5832720    FOODS_3_811_CA_1_validation     1420        2       0         0   
5832721    FOODS_3_812_CA_1_validation     1421        2       0         0   
5832722    FOODS_3_813_CA_1_validation     1422        2       0         0   
5832723    FOODS_3_814_CA_1_validation     1423        2       0         0   
5832724    FOODS_3_815_CA_1_validation     1424        2       0         0   
5832725    FOODS_3_816_CA_1_validation     1425        2       0         0   
5832726    FOODS_3_817_CA_1_validation     1426        2  

In [126]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5832737 entries, 0 to 5832736
Data columns (total 59 columns):
id                      object
item_id                 int64
dept_id                 int64
cat_id                  int64
store_id                int64
state_id                int64
date_block_num          int16
target                  int16
wm_yr_wk                int64
wday                    int16
month                   int16
year                    int64
event_name_1            int64
event_type_1            int64
event_name_2            int64
event_type_2            int64
snap_CA                 int16
snap_TX                 int16
snap_WI                 int16
sell_price              float32
target_store            int16
target_item             int16
target_dept             int16
target_lag_1            float64
target_store_lag_1      float64
target_item_lag_1       float64
target_dept_lag_1       float64
target_lag_2            float64
target_store_lag_2      float64
ta

# Mean encoding

In [127]:
# !!!! The below is not implemented correcly !!!!

# kf = KFold(n_splits=2, shuffle=False)

# for tr_ind, val_ind in kf.split(all_data):
#     X_tr, X_val = all_data.iloc[tr_ind], all_data.iloc[val_ind]
    
#     # Calculate a mapping: {item_id: target_mean}
#     item_id_target_mean = X_tr[(all_data['date_block_num']>=train_end-30) & 
#                                (all_data['date_block_num']<=train_end)].groupby('item_id').target.mean()

#     # Map the computed means to the `item_id`'s
#     all_data.loc[all_data.index[val_ind], 'item_target_enc'] = X_val['item_id'].map(item_id_target_mean)
    
# # # Fill NaNs
# # all_data['item_target_enc'].fillna(0.3343, inplace=True) 

# all_data

In [ ]:
# Given a item_id, we compute mean of the target, and apply it to the next day
# E.g. We compute mean of the target for day 1, and use it as a feature for day 2

# Split into train, val, test sets

We use 30 days as validation data

In [128]:
to_drop_cols = ['id', 'target', 'date_block_num', 'state_id', 'store_id',
                'target_store', 'target_item', 'target_dept']

In [129]:
X_train = all_data[(all_data['date_block_num']>=train_start) & 
                   (all_data['date_block_num']<=train_end)].drop(to_drop_cols, axis = 1)
X_cv = all_data[all_data['date_block_num']==val_date_block_num].drop(to_drop_cols, axis = 1)
# X_train_cv = all_data[all_data['date_block_num']<=1913].drop(to_drop_cols, axis = 1)

y_train = all_data[(all_data['date_block_num']>=train_start) & 
                   (all_data['date_block_num']<=train_end)][['id', 'target']]
y_cv = all_data[all_data['date_block_num']==val_date_block_num][['id', 'target']]
# y_train_cv = all_data[all_data['date_block_num']<=1913][['id', 'target']]

# Train the model and get prediction for day val_date_block_num

In [130]:
# Create the model
model = XGBRegressor(seed=model_seed,
                      n_estimators=n_estimators,
                      max_depth=max_depth,
                      learning_rate=learning_rate,
                      min_child_weight=min_child_weight,
                      objective='reg:squarederror')

# Train the regressor
model.fit(X_train, y_train['target'])

/Users/yibin/anaconda3/lib/python3.7/site-packages/xgboost/core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future version
  if getattr(data, 'base', None) is not None and \
/Users/yibin/anaconda3/lib/python3.7/site-packages/xgboost/core.py:588: FutureWarning: Series.base is deprecated and will be removed in a future version
  data.base is not None and isinstance(data, np.ndarray) \


[07:13:01] WARNING: src/learner.cc:686: Tree method is automatically selected to be 'approx' for faster speed. To use old behavior (exact greedy algorithm on single machine), set tree_method to 'exact'.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
             n_jobs=1, nthread=None, objective='reg:squarederror',
             random_state=0, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
             seed=100, silent=None, subsample=1, verbosity=1)

In [131]:
# View a list of the features and their importance scores
imp = list(zip(X_train.columns, model.feature_importances_))
imp.sort(key=lambda tup: tup[1], reverse=False) 
imp

[('dept_id', 0.0),
 ('cat_id', 0.0),
 ('year', 0.0),
 ('event_name_2', 0.0),
 ('event_type_2', 0.0),
 ('snap_TX', 0.0),
 ('snap_WI', 0.0),
 ('target_item_lag_1', 0.0),
 ('target_item_lag_2', 0.0),
 ('target_item_lag_3', 0.0),
 ('target_dept_lag_3', 0.0),
 ('target_item_lag_4', 0.0),
 ('target_store_lag_5', 0.0),
 ('target_item_lag_5', 0.0),
 ('target_dept_lag_5', 0.0),
 ('target_item_lag_6', 0.0),
 ('target_item_lag_7', 0.0),
 ('target_item_lag_30', 0.0),
 ('target_dept_lag_30', 0.0),
 ('target_store_lag_365', 0.0),
 ('target_item_lag_365', 0.0),
 ('target_dept_lag_365', 0.0),
 ('target_store_lag_4', 0.00047523298),
 ('event_type_1', 0.0005568701),
 ('target_dept_lag_4', 0.0008178162),
 ('event_name_1', 0.001369444),
 ('target_dept_lag_1', 0.0015169038),
 ('snap_CA', 0.0015522038),
 ('target_dept_lag_6', 0.0017914154),
 ('target_store_lag_30', 0.0020111254),
 ('target_store_lag_6', 0.002414235),
 ('wm_yr_wk', 0.0028687438),
 ('target_dept_lag_7', 0.00328397),
 ('target_dept_lag_2', 0.0

In [132]:
# Keep only features that have a non-zero importance score
features = [
    'sell_price',
    'target_lag_365',
    'target_dept_lag_1',
    'item_id',
    'target_lag_30',
    'target_lag_6',
    'target_lag_5',
    'target_lag_4',
    'target_lag_2',
    'target_lag_3',
    'target_lag_7',
    'target_lag_1'
]

# Retrain model
model.fit(X_train[features], y_train['target'])

[08:20:34] WARNING: src/learner.cc:686: Tree method is automatically selected to be 'approx' for faster speed. To use old behavior (exact greedy algorithm on single machine), set tree_method to 'exact'.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
             n_jobs=1, nthread=None, objective='reg:squarederror',
             random_state=0, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
             seed=100, silent=None, subsample=1, verbosity=1)

In [133]:
# Do prediction
pred = model.predict(X_cv[features]).astype('int')

# Calculate RMSE
print("MSE = " + str(mean_squared_error(y_cv['target'], pred)))

MSE = 4.888488028861922


In [134]:
y_cv['pred'] = pred
y_cv

id  target  pred
5747365  HOBBIES_1_001_CA_1_validation       1     1
5747366  HOBBIES_1_002_CA_1_validation       1     0
5747367  HOBBIES_1_003_CA_1_validation       0     0
5747368  HOBBIES_1_004_CA_1_validation       0     3
5747369  HOBBIES_1_005_CA_1_validation       1     0
5747370  HOBBIES_1_006_CA_1_validation       0     1
5747371  HOBBIES_1_007_CA_1_validation       1     0
5747372  HOBBIES_1_008_CA_1_validation       0     3
5747373  HOBBIES_1_009_CA_1_validation       0     0
5747374  HOBBIES_1_010_CA_1_validation       0     0
5747375  HOBBIES_1_011_CA_1_validation       0     0
5747376  HOBBIES_1_012_CA_1_validation       1     0
5747377  HOBBIES_1_013_CA_1_validation       0     0
5747378  HOBBIES_1_014_CA_1_validation       3     1
5747379  HOBBIES_1_015_CA_1_validation       4     2
5747380  HOBBIES_1_016_CA_1_validation       2     5
5747381  HOBBIES_1_017_CA_1_validation       0     0
5747382  HOBBIES_1_018_CA_1_validation       0     0
5747383  HOBBIES_1_019_CA_1_validation       5     4
5747384  HOBBIES_1_020_CA_1_validation       0     0
5747385  HOBBIES_1_021_CA_1_validation       0     1
5747386  HOBBIES_1_022_CA_1_validation       1     0
5747387  HOBBIES_1_023_CA_1_validation       1     1
5747388  HOBBIES_1_024_CA_1_validation       0     0
5747389  HOBBIES_1_025_CA_1_validation       0     0
5747390  HOBBIES_1_026_CA_1_validation       0     0
5747391  HOBBIES_1_027_CA_1_validation       1     0
5747392  HOBBIES_1_028_CA_1_validation       1     0
5747393  HOBBIES_1_029_CA_1_validation       2     2
5747394  HOBBIES_1_030_CA_1_validation      10     8
...                                ...     ...   ...
5750384    FOODS_3_798_CA_1_validation       0     0
5750385    FOODS_3_799_CA_1_validation       0     0
5750386    FOODS_3_800_CA_1_validation       3     2
5750387    FOODS_3_801_CA_1_validation       2     2
5750388    FOODS_3_802_CA_1_validation       0     0
5750389    FOODS_3_803_CA_1_validation       1     2
5750390    FOODS_3_804_CA_1_validation       1     2
5750391    FOODS_3_805_CA_1_validation       1     1
5750392    FOODS_3_806_CA_1_validation       1     0
5750393    FOODS_3_807_CA_1_validation      11     2
5750394    FOODS_3_808_CA_1_validation      16     4
5750395    FOODS_3_809_CA_1_validation       4     2
5750396    FOODS_3_810_CA_1_validation       8     5
5750397    FOODS_3_811_CA_1_validation       9     8
5750398    FOODS_3_812_CA_1_validation       2     2
5750399    FOODS_3_813_CA_1_validation       1     2
5750400    FOODS_3_814_CA_1_validation       2     1
5750401    FOODS_3_815_CA_1_validation       0     0
5750402    FOODS_3_816_CA_1_validation       8     3
5750403    FOODS_3_817_CA_1_validation       1     1
5750404    FOODS_3_818_CA_1_validation       4     3
5750405    FOODS_3_819_CA_1_validation       2     2
5750406    FOODS_3_820_CA_1_validation       1     2
5750407    FOODS_3_821_CA_1_validation       4     2
5750408    FOODS_3_822_CA_1_validation       2     2
5750409    FOODS_3_823_CA_1_validation       0     0
5750410    FOODS_3_824_CA_1_validation       0     0
5750411    FOODS_3_825_CA_1_validation       1     1
5750412    FOODS_3_826_CA_1_validation       0     1
5750413    FOODS_3_827_CA_1_validation       3     4

[3049 rows x 3 columns]

# Get predictions for validation set 28 days

In [135]:
# Rename columns
sale_tr_val_melt = sale_tr_val_melt.rename(columns={'variable': 'date_block_num', 
                                                    'value': 'target'})

In [136]:
# Convert variable to a numerical column ie. d_1 to 1
sale_tr_val_melt['date_block_num'] = sale_tr_val_melt['date_block_num'].str.split('_').str[1].astype('int16')
sale_tr_val_melt.tail()

id      item_id  dept_id cat_id store_id  \
5832732  FOODS_3_823_CA_1_validation  FOODS_3_823  FOODS_3  FOODS     CA_1   
5832733  FOODS_3_824_CA_1_validation  FOODS_3_824  FOODS_3  FOODS     CA_1   
5832734  FOODS_3_825_CA_1_validation  FOODS_3_825  FOODS_3  FOODS     CA_1   
5832735  FOODS_3_826_CA_1_validation  FOODS_3_826  FOODS_3  FOODS     CA_1   
5832736  FOODS_3_827_CA_1_validation  FOODS_3_827  FOODS_3  FOODS     CA_1   

        state_id  date_block_num  target  
5832732       CA            1913       1  
5832733       CA            1913       0  
5832734       CA            1913       2  
5832735       CA            1913       0  
5832736       CA            1913      21

In [137]:
# Convert variable to a numerical column ie. d_1 to 1
calendar['d'] = calendar['d'].str.split('_').str[1].astype('int16')
calendar.head()

date  wm_yr_wk  wday  month  year  d event_name_1 event_type_1  \
0  2011-01-29     11101     1      1  2011  1          nil          nil   
1  2011-01-30     11101     2      1  2011  2          nil          nil   
2  2011-01-31     11101     3      1  2011  3          nil          nil   
3  2011-02-01     11101     4      2  2011  4          nil          nil   
4  2011-02-02     11101     5      2  2011  5          nil          nil   

  event_name_2 event_type_2  snap_CA  snap_TX  snap_WI  
0          nil          nil        0        0        0  
1          nil          nil        0        0        0  
2          nil          nil        0        0        0  
3          nil          nil        1        1        0  
4          nil          nil        1        0        1

In [138]:
# Do prediction on the test set
results = y_cv[['id']].copy()
sale_tr_val_melt2 = sale_tr_val_melt[sale_tr_val_melt['date_block_num']<pred_start]
mse = []

for i in range(pred_start, pred_end+1):
    print("\nday = " + str(i))
    
    # Create the test set here
    temp = sale_tr_val_melt2[sale_tr_val_melt2['date_block_num']==1].copy()
    temp.loc[:, 'target'] = 0
    temp.loc[:, 'date_block_num'] = i
    sale_tr_val_melt2 = sale_tr_val_melt2.append(temp)
    
    # Merge with calendar to get events info
    sale_tr_val_melt_merged = sale_tr_val_melt2.merge(calendar, 
                                                     left_on='date_block_num', 
                                                     right_on='d', 
                                                     how='left')

    # Drop unneccesary columns
    sale_tr_val_melt_merged.drop(columns=['d'], inplace=True)
    
    # Merge with sell_prices to get price info
    all_data = sale_tr_val_melt_merged.merge(sell_prices, 
                                             left_on=['store_id', 'item_id', 'wm_yr_wk'], 
                                             right_on=['store_id', 'item_id', 'wm_yr_wk'], 
                                             how='left')

    # Remove unnecessaray column
    all_data.drop(['date'], axis=1, inplace=True)
    
    # Convert features to categorical 
    for column in columns_to_be_labeled:
        all_data.loc[:, column] = LabelEncoder().fit_transform(all_data[column]) # fit_transform: Fit label encoder and return encoded labels. Note the encoded labels are integers!!

    # Create features
    all_data = get_agg_features(all_data)
       
    # Get the lag features
    all_data = get_lags(index_cols, cols_to_rename, all_data, shift_range)
    
    # Get the test set
    X_test = all_data[all_data['date_block_num']==i].drop(to_drop_cols, axis = 1)
    y_test = all_data[all_data['date_block_num']==i][['id', 'target']]
    
    # Do prediction on test set
    pred = model.predict(X_test[features]).astype('int')
    print("No. of zeros = " + str(len(pred[pred==0])))
    
    # Add preds to dataset
    sale_tr_val_melt2.loc[sale_tr_val_melt2['date_block_num']==i, 'target'] = pred 

    # Calculate RMSE
    print("MSE = " + str(mean_squared_error(y_test['target'], pred)))
    mse.append(mean_squared_error(y_test['target'], pred))
    
    results['F'+str(i-pred_start+1)] = pred
    
print("mse mean = " + str(np.mean(mse)))
results


day = 1886
No. of zeros = 1809
MSE = 10.751065923253526

day = 1887
No. of zeros = 1897
MSE = 8.777304034109545

day = 1888
No. of zeros = 1971
MSE = 7.397835355854379

day = 1889
No. of zeros = 2016
MSE = 8.187602492620531

day = 1890
No. of zeros = 2042
MSE = 11.011151197113808

day = 1891
No. of zeros = 2086
MSE = 13.848474909806495

day = 1892
No. of zeros = 2175
MSE = 14.311577566415218

day = 1893
No. of zeros = 2231
MSE = 11.339127582814037

day = 1894
No. of zeros = 2292
MSE = 9.230895375532961

day = 1895
No. of zeros = 2372
MSE = 8.354542472941947

day = 1896
No. of zeros = 2421
MSE = 9.486061003607741

day = 1897
No. of zeros = 2458
MSE = 10.47294194817973

day = 1898
No. of zeros = 2474
MSE = 12.244670383732371

day = 1899
No. of zeros = 2500
MSE = 12.747458183010822

day = 1900
No. of zeros = 2519
MSE = 11.711380780583799

day = 1901
No. of zeros = 2539
MSE = 10.195473925877337

day = 1902
No. of zeros = 2562
MSE = 9.395867497540177

day = 1903
No. of zeros = 2580
MSE = 9

id  F1  F2  F3  F4  F5  F6  F7  F8  F9  \
5747365  HOBBIES_1_001_CA_1_validation   1   1   0   0   0   0   0   0   0   
5747366  HOBBIES_1_002_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747367  HOBBIES_1_003_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747368  HOBBIES_1_004_CA_1_validation   3   2   2   2   2   2   2   2   1   
5747369  HOBBIES_1_005_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747370  HOBBIES_1_006_CA_1_validation   1   1   1   1   1   1   1   1   1   
5747371  HOBBIES_1_007_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747372  HOBBIES_1_008_CA_1_validation   3   5   4   2   2   2   3   3   4   
5747373  HOBBIES_1_009_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747374  HOBBIES_1_010_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747375  HOBBIES_1_011_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747376  HOBBIES_1_012_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747377  HOBBIES_1_013_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747378  HOBBIES_1_014_CA_1_validation   1   1   1   1   1   1   1   0   1   
5747379  HOBBIES_1_015_CA_1_validation   2   2   2   2   2   2   3   2   2   
5747380  HOBBIES_1_016_CA_1_validation   5   6   6   7   6   7   6   5   5   
5747381  HOBBIES_1_017_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747382  HOBBIES_1_018_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747383  HOBBIES_1_019_CA_1_validation   4   4   5   4   4   3   3   3   3   
5747384  HOBBIES_1_020_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747385  HOBBIES_1_021_CA_1_validation   1   1   1   1   0   0   0   0   0   
5747386  HOBBIES_1_022_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747387  HOBBIES_1_023_CA_1_validation   1   0   0   0   0   0   0   0   0   
5747388  HOBBIES_1_024_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747389  HOBBIES_1_025_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747390  HOBBIES_1_026_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747391  HOBBIES_1_027_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747392  HOBBIES_1_028_CA_1_validation   0   0   0   0   0   0   0   0   0   
5747393  HOBBIES_1_029_CA_1_validation   2   1   1   1   1   1   1   1   1   
5747394  HOBBIES_1_030_CA_1_validation   8   6   5   6   7   7   8   7   6   
...                                ...  ..  ..  ..  ..  ..  ..  ..  ..  ..   
5750384    FOODS_3_798_CA_1_validation   0   0   0   0   0   0   0   0   0   
5750385    FOODS_3_799_CA_1_validation   0   0   0   0   0   0   0   0   0   
5750386    FOODS_3_800_CA_1_validation   2   2   2   2   2   2   1   1   1   
5750387    FOODS_3_801_CA_1_validation   2   1   1   1   1   1   1   1   1   
5750388    FOODS_3_802_CA_1_validation   0   0   0   0   0   0   0   0   0   
5750389    FOODS_3_803_CA_1_validation   2   2   2   2   2   2   2   2   2   
5750390    FOODS_3_804_CA_1_validation   2   3   2   2   3   2   2   2   2   
5750391    FOODS_3_805_CA_1_validation   1   1   1   1   0   0   0   0   0   
5750392    FOODS_3_806_CA_1_validation   0   0   0   0   0   0   0   0   0   
5750393    FOODS_3_807_CA_1_validation   2   2   3   2   2   2   2   2   2   
5750394    FOODS_3_808_CA_1_validation   4   4   4   4   4   5   4   3   3   
5750395    FOODS_3_809_CA_1_validation   2   2   2   2   2   2   1   1   1   
5750396    FOODS_3_810_CA_1_validation   5   4   3   3   4   4   3   3   2   
5750397    FOODS_3_811_CA_1_validation   8   8   9   8   9  10  10   9   9   
5750398    FOODS_3_812_CA_1_validation   2   2   2   1   2   1   1   1   1   
5750399    FOODS_3_813_CA_1_validation   2   2   2   2   2   1   1   1   1   
5750400    FOODS_3_814_CA_1_validation   1   1   1   1   1   0   0   0   0   
5750401    FOODS_3_815_CA_1_validation   0   0   0   0   0   0   0   0   0   
5750402    FOODS_3_816_CA_1_validation   3   3   3   3   4   3   3   4   3   
5750403    FOODS_3_817_CA_1_validation   1   0   0   0   0

In [139]:
# Store results as csv
results.to_csv("./out/v4_results-" + state + "-" + store + "-" + str(pred_start) + "-" + str(pred_end) + ".csv", index=False)

In [140]:
tocEnd = time.time()
print("Total time = " + str((tocEnd-ticStart)/60.0) + " mins")

Total time = 131.44092045227686 mins


# Findings

Without hyperparam tuning, state='CA', store_id='CA_1':
* If train on 1885, pred on 1886 to 1913, mean mse = 12.47
* If train on <=1885, pred on 1886 to 1913, mean mse = 10.51

# Combine all predictions together and prepare submission file

In [141]:
# # Combine all predictions together
# pred_start = 1914              # do pred starting from this day
# pred_end = 1941                # do pred until and include this day

# state_store_list = [('CA', 'CA_1'), ('CA', 'CA_2'), ('CA', 'CA_3'), ('CA', 'CA_4'),
#                     ('TX', 'TX_1'), ('TX', 'TX_2'), ('TX', 'TX_3'), 
#                     ('WI', 'WI_1'), ('WI', 'WI_2'), ('WI', 'WI_3')]

# state_store = state_store_list[0]
# results_tot = pd.read_csv("./out/results-" + state_store[0] + "-" + state_store[1] + "-" + str(pred_start) + "-" + str(pred_end) + ".csv")


# for state_store in state_store_list[1:]:
#     results = pd.read_csv("./out/v4_results-" + state_store[0] + "-" + state_store[1] + "-" + str(pred_start) + "-" + str(pred_end) + ".csv")
#     results_tot = results_tot.append(results)
    
# results_tot

In [142]:
# # Prepare submission file
# submission = pd.merge(sample_sub[['id']], results_tot, on='id', how='left')
# submission.fillna(value=0, inplace=True)
# submission.to_csv("./out/v4_submission.csv", index=False)
# submission

This obtained weighted RMSSE of ? on the validation set